compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,483 @@
+    }
+
+    new_shape_idx++;
+    move_to_next_range();
+  }
+  _shape = new_shape;
+  return true;
+}
+
+bool TaggedShapeAnalyzer::verify_tag() const
+{
+  // check whether tags in _shape are incremental
+  uint8_t tag = _start_tag;
+  for (const auto &dim : _shape)
+  {
+    for (const auto &t : dim.tags)
+    {
+      if (t == tag)
+        tag++;
+      else
+        return false;
+    }
+  }
+  return true;
+}
+
+bool TaggedShapeAnalyzer::can_remove_transposes(const luci::CircleTranspose *f_tr,
+                                                const luci::CircleReshape *m_rs,
+                                                const luci::CircleTranspose *b_tr)
+{
+  // handle only perm with CicleConst(S32, S64) type
+  auto analyze_transpose_by_datatype = [this](const luci::CircleTranspose *tr) -> bool {
+    bool res = false;
+
+    const auto perm = dynamic_cast<luci::CircleConst *>(tr->perm());
+    if (perm == nullptr)
+      return res;
+
+    switch (perm->dtype())
+    {
+      case loco::DataType::S32:
+        res = analyze_transpose<loco::DataType::S32>(tr);
+        break;
+      case loco::DataType::S64:
+        res = analyze_transpose<loco::DataType::S64>(tr);
+        break;
+      default:
+        break;
+    }
+    return res;
+  };
+
+  // handle only shape with CirlceCosnt(S32, S64) type
+  auto analyze_reshape_by_datatype = [this](const luci::CircleReshape *rs) -> bool {
+    bool res = false;
+
+    const auto shape = dynamic_cast<luci::CircleConst *>(rs->shape());
+    if (shape == nullptr)
+      return res;
+
+    switch (shape->dtype())
+    {
+      case loco::DataType::S32:
+        res = analyze_reshape<loco::DataType::S32>(rs);
+        break;
+      case loco::DataType::S64:
+        res = analyze_reshape<loco::DataType::S64>(rs);
+        break;
+      default:
+        break;
+    }
+    return res;
+  };
+
+  const luci::CircleNode *in_tensor = loco::must_cast<luci::CircleNode *>(f_tr->a());
+
+  init_shape_with_tag(in_tensor);
+
+  if (analyze_transpose_by_datatype(f_tr) == false)
+    return false;
+
+  if (analyze_reshape_by_datatype(m_rs) == false)
+    return false;
+
+  if (analyze_transpose_by_datatype(b_tr) == false)
+    return false;
+
+  if (verify_tag() == false)
+    return false;
+
+  return true;
+}
+
+/**
+ * @brief create CircleReshape node that reshapes 'front_transpose input tensor shape' into
+ * 'back_transposes output tensor shape'
+ */
+template <loco::DataType ShapeType>
+luci::CircleReshape *create_reshape_node(loco::Graph *graph,
+                                         const luci::CircleTranspose *front_transpose,
+                                         const luci::CircleReshape *middle_reshape,
+                                         const luci::CircleTranspose *back_transpose)
+{
+  std::string composed_name =
+    front_transpose->name() + "/" + middle_reshape->name() + "/" + back_transpose->name();
```

Comment:
from naming rules of TensorFlow, which we(?) follow, it would be better to use `;` instead of `/`